### DISCUSSION TITLE
Proposal: Quadruple precision floats

### DISCUSSION THREAD

**Discussion by TheUnlocked** [dotnet/csharplang#8199] [Language Ideas]
As the title suggests, adding a new `quad` type with a literal like `3.141592653589793238462q`. I understand that `Decimal` currently exists, but it has a very small maximum range compared to the massive range of a quadruple precision float (or even just a `double`). I suppose that a `qint` and `uqint` could also be added as well, though probably with better thought out names.

A potential drawback is that quadruple precision math doesn't tend to have the same kind of hardware acceleration that double precision math does, but sometimes the extra precision is worth that price.

**Comment by tannergooding** [comment]
This would require a corresponding type in CoreFX first.

Additionally, given that IL doesn't support literals for non-primitive types, there would need to be an alternative mechanism for supporting `quad` constants. Such support is probably better suited to a more general language feature (such as `blittable constants`).

### As a workaround
.NET Core provides the `BigNumber` type which supports an arbitrary precision: https://source.dot.net/#System.Runtime.Numerics/System/Numerics/BigNumber.cs,969928e529663ace

**Comment by TheUnlocked** [comment]
Ah, I wasn't aware that .NET Core had a bignum implementation. I'll leave this open anyways because it still might be nice to have 128-bit numbers.

**Comment by iam3yal** [comment]
@TheUnlocked The question is what would be the benefit of 128-bit number as part of the language? I mean, "nice to have" can be interpreted as "redundant/unnecessary" so you may want to cover how the language benefits from this.

**Comment by TheUnlocked** [comment]
I initially thought that this would be useful upon seeing how the game Kerbal Space Program (written in C#) starts to bug out when you have a massive number of objects interacting with eachother due to double imprecision. While I'm not suggesting that it would make any sense to add a new language feature for a single program, that particular example shows that when dealing with massive simulations with a lot of moving pieces, doubles sometimes aren't enough.

I suppose another interesting thing to see would be a benchmarking (probably written in a language that does support quadruple precision) between quads and BigNumbers, to see if quads are sufficiently faster and less memory intensive to be worth it.

**Comment by 4creators** [comment]
> The question is what would be the benefit of 128-bit number as part of the language?

@eyalsk 

This proposal is about implementing one of the binary formats from IEEE 754:2008 standard (currently translated into ISO/IEC/IEEE 60559:2011). In general implementation of parts or all of the IEEE 754:2008 is tracked by several issues in coreclr and corefx repos. Language support would be helpful at the level of `numeric literals` supporting new types initialization or constant definition and perhaps, however not strictly required for having this type available in framework, addition of new built in types to support new numeric types (i.e. `float16`, `float128`).

![ieee754-2008_1](https://user-images.githubusercontent.com/801959/34910223-e2ab8df6-f8b0-11e7-8048-e4710209fb4b.PNG)

![ieee754-2008_2](https://user-images.githubusercontent.com/801959/34910380-0d6bfabe-f8b4-11e7-891a-3ff7d1b932ab.PNG)

**Comment by IS4Code** [comment]
It depends on the relative performance of the type to other floating-point types. If `float128` was a thing understood at least by the JITter, it could be optimized to rely on the features of the CPU and it would make sense to include it in the list of type keywords, but not if it would be just another `decimal`.

Language support in this case either means special understanding of operators and emitting a special signature, but that means modifying CIL to support `float128`, which isn't probably going to happen. So it would be just something like mapping `quad` to `System.Quadruple` and understanding `const quad`.

**Comment by Happypig375** [comment]
Cross-reference to corresponding issue at corefx: https://github.com/dotnet/corefx/issues/17267

**Comment by asyncawaitmvvm** [comment]
If you are creating big number types without matching trigonometry/log functions, please have a seat. Don't waste your time. Only a complete implementation is useful, otherwise it's a high school project.

> @TheUnlocked The question is what would be the benefit of 128-bit number as part of the language? I mean, "nice to have" can be interpreted as "redundant/unnecessary" so you may want to cover how the language benefits from this.

IPv6 addresses and Guid literals, derp. Quicker static initialization. Combine this with something like constexpr and it would be just lordly.

**Comment by MovGP0** [comment]
There is only very few hardware (source: [Wikipedia](https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format#Hardware_support)) that supports quadruple precision floats.

As such, (most of) the calculations would have to be implemented in software and very slow.

An dedicated library would be probably a better fit for such a type than it beeing an integral part of the framework.

**Reply by jnm2** [reply]
Though, it wouldn't be an integral part of the framework, because where would be the point?

**Comment by migueldecimal128** [comment]
I am putting this here because decimal128 was mentioned ...

I have implemented IEEE 754-2019 decimal128 for C# + C/Swift/Rust/Go/Zig/Java/Kotlin Multiplatform/Python. 
Native implementations, not a wrapper around C ... except for Python. 
All flags and all roundingDirections per the spec. 
Includes exponential functions ln, exp, log10, exp10 ... which enable raise to a fractional power. 
Passes 3 known sets of industry test vectors to verify conformance. 

https://decimal128.com
https://github.com/migueldecimal128/decimal128-csharp

I'm not really a C# developer (this low-level implementation is basically C code), so I need assistance adapting the API to C# idioms and bringing this to the C# community.

**Reply by huoyaoyuan** [reply]
Decimal128 has just been added to the BCL recently (https://github.com/dotnet/runtime/pull/100729). It will be available in .NET 11. It's implemented in pure C#.

**Reply by migueldecimal128** [reply]
I was not aware fo this ... thanks for the pointer

**Reply by huoyaoyuan** [reply]
We can all complete IEEE Decimal arithmetic without the assistance of AI. The code quality of AI is far from acceptable.

**Reply by migueldecimal128** [reply]
*you folks* can do it for C# without AI. But, all by myself I can't port to 8 other programming languages without AI. :)

**Reply by tannergooding** [reply]
Notably, the C# implementation wasn't done "without AI". I explicitly used it and it helped simplify the overall process and ensure that I could land the remaining massive amount of supporting code quickly.

However, I also used it as an "expert" in this space and with the existing knowledge/reference of how to implement it "correctly". I also used a well known and open source (BSD 3-clause) reference implementation (Intel DFP) with appropriate 3rd party attribution and some static tooling (my own `ClangSharp P/Invoke Generator`) to do the baseline porting.

So the use of AI was primarily in helping do more complete testing of the data beyond the well known oracle test beds, helping ensure that I made consistent changes when doing the port, handling C to C# translations that were less trivial (but still manually audited and checked by myself), and ensuring the targeted optimizations that I was making were consistently applied across the ported code.

It wasn't used to for example "vibe code" the implementation into C# or ignore the appropriate levels of code review and tuning that I would've already been doing by hand. It was just used to accelerate what I was already doing alongside existing static tools and development processes, to help make it all the more robust.

**Reply by migueldecimal128** [reply]
For my part, I hand-coded implementations in Kotlin and Swift, focusing on correctness and performance, esp division by powers of 10. I then used AI to enforce a relatively portable language subset (essentially C, functions, not methods) across other ports. Architecture is exactly the same across all ports. Function names are the same across all ports. I also found AI very helpful with QA, esp building harnesses for existing (text file) test vectors.
